### PR TITLE
Write an event for a decision evaluation

### LINKED ISSUE #8115: Write an event for a decision evaluation
**Description**

We should write an event when a decision is evaluated. The event should include all data about the decision evaluation (e.g. the decision result, the result of required decisions, etc.). The event can be consumed to analyze and visualize the evaluation of the decision, for example, by Operate and Optimize. 

* create a new record `DecisionEvaluationRecord` 
* create a new intent `evaluated`
* it should include all data about the evaluation that is provided from the DMN module 
* it should include context data about the process instance that the decision is called from    

The record could look like this:

```
{
  "decisionKey": 1L,
  "decisionId": "x",
  "version": 1L,
  "decisionRequirementsKey": 1L,
  "decisionRequirementsId": "x",
  "processDefinitionKey": 1L,
  "processInstanceKey": 1L,
  "elementInstanceKey": 1L,
  "elementId": "x",
  "decisionOutput": "x" (JSON),
  "evaluatedDecisions": [
    {
      "decisionId": "x",
      "decisionName": "x",
      "decisionType": "x",
      "decisionOutput": "x" (JSON),  
      "evaluatedInputs": [ (for decision tables) 
      	{
      	  "inputId": "x",
      	  "inputName": "x",
      	  "inputValue": "x" (JSON) 
      	}
      ],
      "matchedRules": [ (for decision tables)
      	{
      	  "ruleId": "x",
      	  "ruleIndex:": x,
      	  "evaluatedOutputs": [
  	    {
  	      "outputId": "x",
  	      "outputName": "x",
  	      "outputValue": "x" (JSON)
  	    } 
      	  ] 
      	}
      ]
    }
  ]  
}
```

blocked by #8080 
blocked by #8112

### PR DESCRIPTION
**Pull request by saig0**
## Description

* add new value type `DECISION_EVALUATION`
* add a new record for a decision evaluation
* write the new record when a decision is evaluated

## Related issues

closes #8115 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/BusinessRuleTaskTest.java [hunk 1]
```diff
@@ -205,4 +209,129 @@ public void shouldUseResultInOutputMappings() {
...
+                processWithBusinessRuleTask(
+                    t ->
+                        t.zeebeCalledDecisionId("force_user").zeebeResultVariable(RESULT_VARIABLE)))
+            .deploy();
+
+    final var calledDecision =
+        deployment.getValue().getDecisionsMetadata().stream()
+            .filter(decision -> decision.getDecisionId().equals("force_user"))
+            .findFirst()
+            .orElseThrow();
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/BusinessRuleTaskTest.java:230]: 👍 Often we do a `getFirst()` I think, but I like this better. It's nice and explicit

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/BusinessRuleTaskTest.java [hunk 2]
```diff
@@ -205,4 +209,129 @@ public void shouldUseResultInOutputMappings() {
...
+    final Map<String, Object> variables =
+        Map.ofEntries(entry("lightsaberColor", "blue"), entry("height", 182));
+    final long processInstanceKey =
+        ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).withVariables(variables).create();
+
+    final var businessRuleTaskActivated =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .withElementType(BpmnElementType.BUSINESS_RULE_TASK)
+            .getFirst();
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/BusinessRuleTaskTest.java:242]: ❓ Why not continue using `findFirst().orElseThrow()`?

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java [hunk 3]
```diff
@@ -70,27 +95,18 @@ public BpmnDecisionBehavior(
...
     // related incidents.
     resultOrFailure.ifRight(
-        result ->
-            triggerProcessEventWithResultVariable(context, element.getResultVariable(), result));
+        result -> {
+          triggerProcessEventWithResultVariable(context, element.getResultVariable(), result);
+
+          final var decision = decisionOrFailure.get();
+          writeDecisionEvaluationEvent(decision, result, context);
+        });
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java:103]: 🔧 This change obfuscates what is meant by the comment block above. Please put writing the decisionEvaluationEvent into its own call to `resultOrFailure.ifRight()`. That way its more clear that the comment is only in relation to the triggerProcessEventWithResultVariable

**Review by korthout — APPROVED** [review]
Thanks @saig0 LGTM 🚀 

I've only added some minor comments, but nothing blocking

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/BusinessRuleTaskTest.java (hunk 2 above)
Comment (saig0) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/BusinessRuleTaskTest.java:242]: Different context. `findFirst().orElseThrow()` is used on a Java stream that is created from a deployment event but not from a `RecordStream`.

I agree that `findFirst().orElseThrow()` is in general more explicit but we already use `getFirst()` in other places :sweat_smile:

**Comment by saig0** [comment]
bors merge

**Pull request merged**